clang-tools-extra/clang-tidy/portability/AvoidPlatformSpecificFundamentalTypesCheck.cpp

```diff
@@ -0,0 +1,183 @@
+      WarnOnFloats(Options.get("WarnOnFloats", true)),
+      WarnOnInts(Options.get("WarnOnInts", true)),
+      WarnOnChars(Options.get("WarnOnChars", true)),
+      IncludeInserter(Options.getLocalOrGlobal("IncludeStyle",
+                                               utils::IncludeSorter::IS_LLVM),
+                      areDiagsSelfContained()) {
+  if (!WarnOnFloats && !WarnOnInts && !WarnOnChars)
+    this->configurationDiag(
+        "The check 'portability-avoid-platform-specific-fundamental-types' "
+        "will not perform any analysis because 'WarnOnFloats', 'WarnOnInts' "
+        "and 'WarnOnChars' are all false.");
+}
+
+void AvoidPlatformSpecificFundamentalTypesCheck::registerPPCallbacks(
+    const SourceManager &SM, Preprocessor *PP, Preprocessor *ModuleExpanderPP) {
+  IncludeInserter.registerPreprocessor(PP);
+}
+
+void AvoidPlatformSpecificFundamentalTypesCheck::storeOptions(
+    ClangTidyOptions::OptionMap &Opts) {
+  Options.store(Opts, "WarnOnFloats", WarnOnFloats);
+  Options.store(Opts, "WarnOnInts", WarnOnInts);
+  Options.store(Opts, "WarnOnChars", WarnOnChars);
+  Options.store(Opts, "IncludeStyle", IncludeInserter.getStyle());
+}
+
+static std::string getFloatReplacement(const BuiltinType *BT,
+                                       ASTContext &Context) {
+  const TargetInfo &Target = Context.getTargetInfo();
+
+  auto GetReplacementType = [](unsigned Width) {
+    switch (Width) {
+    // This is ambiguous by default since it could be bfloat16 or float16
+    case 16U:
+      return "";
+    case 32U:
+      return "float32_t";
+    case 64U:
+      return "float64_t";
+    case 128U:
+      return "float128_t";
+    default:
+      return "";
+    }
+  };
+
+  switch (BT->getKind()) {
+  // Not an ambiguous type
+  case BuiltinType::BFloat16:
+    return "bfloat16_t";
+  case BuiltinType::Half:
+    return GetReplacementType(Target.getHalfWidth());
+  case BuiltinType::Float:
+    return GetReplacementType(Target.getFloatWidth());
+  case BuiltinType::Double:
+    return GetReplacementType(Target.getDoubleWidth());
+  default:
+    return "";
+  }
+}
+
+void AvoidPlatformSpecificFundamentalTypesCheck::registerMatchers(
+    MatchFinder *Finder) {
+  auto PlatformSpecificFundamentalType = qualType(allOf(
+      builtinType(), anyOf(WarnOnInts ? isBuiltinInt() : unless(anything()),
+                           WarnOnFloats ? isBuiltinFloat() : unless(anything()),
+                           WarnOnChars ? isChar() : unless(anything()),
+                           WarnOnChars ? isWideChar() : unless(anything()))));
+
+  if (!WarnOnInts && !WarnOnFloats && !WarnOnChars)
+    return;
+
+  Finder->addMatcher(typeLoc(loc(PlatformSpecificFundamentalType)).bind("type"),
+                     this);
+}
+
+void AvoidPlatformSpecificFundamentalTypesCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *TL = Result.Nodes.getNodeAs<TypeLoc>("type");
+  if (!TL)
+    return;
```

Comment:
```suggestion
  assert(TL);
```
We expect matchers to work, https://llvm.org/docs/CodingStandards.html#assert-liberally